### PR TITLE
[component] Remove ReportStatus from component.TelemetrySettings

### PR DESCRIPTION
**Pull request by TylerHelmuth**
#### Description

This PR removes `ReportStatus` from `component.TelemetrySettings` and instead expects components to check if their `component.Host` implements a new `componentstatus.Reporter` interface.

#### Link to tracking issue
Related to https://github.com/open-telemetry/opentelemetry-collector/pull/10725
Related to https://github.com/open-telemetry/opentelemetry-collector/pull/10413

#### Testing
unit tests and a sharedinstance e2e test.

The contrib tests will fail because this is a breaking change.  If we merge this I and @mwear can commit to updating contrib before the next release.

### CODE REVIEW

File: component/telemetry.go
Comment (TylerHelmuth) [reply component/telemetry.go:38]: I've changed my mind.  Keeping the `component` stuff around is very messy bc it expects to be using its types instead of `componentstatus`.  So `TelemetrySettings.ReportStatus` would use `component.StatusEvent` but `componentstatus.Reporter` uses `componentstatus.Event`.

I'll commit to updating jaeger once this breaking change is merged.  Luckily it is a one line change to `componentstatus.ReportStatus`, like in the otlpreceiver.

Comment (TylerHelmuth) [reply component/telemetry.go:38]: @yuriolisa according to some Github Searches jaeger is the only external consumer of this API.  This PR is going to break:

- https://github.com/jaegertracing/jaeger/blob/main/pkg/telemetery/settings.go
- https://github.com/jaegertracing/jaeger/blob/main/cmd/collector/app/handler/otlp_receiver.go
- https://github.com/jaegertracing/jaeger/blob/main/cmd/remote-storage/app/server.go
- https://github.com/jaegertracing/jaeger/blob/main/cmd/jaeger/internal/integration/storagecleaner/extension.go
- https://github.com/jaegertracing/jaeger/blob/main/cmd/jaeger/internal/extension/remotesampling/extension.go
- https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/server.go

The path forward for all scenarios is straightforward and I can submit the PR. The only interesting change would be to https://github.com/jaegertracing/jaeger/blob/main/pkg/telemetery/settings.go, as it would be a type change on a public field on the public struct. It looks like that struct was added very recently, so will changing it be ok? The alternative is to deprecate its `ReportStatus` and add a new one that uses `componentstatus.Event`.

File: internal/sharedcomponent/sharedcomponent.go
Comment (mwear) [inline internal/sharedcomponent/sharedcomponent.go:129]: Not an actionable comment, but I just wanted to highlight that managing multiple sources would be unnecessary if we improved our component instancing story as suggested here: https://github.com/open-telemetry/opentelemetry-collector/issues/10534. Another thing to mention, but also likely unactionable, is that we have two options for reporting status from within the shared component. One is to treat the shared component as multiple instances, one per pipeline. That is the approach taken here. The other, would be to treat it as a single instance in multiple pipelines. The latter is more correct, but there are technical difficulties in achieving that. Part of the reason is that the service graph treats the shared component as multiple instances.

File: service/service.go
Comment (mwear) [inline service/service.go:199]: I think we can remove `Ready()` and the ready related code in internal/status. When `ReportStatus` was on `TelemetrySettings` it was possible to call it before start. Now that it's been moved to `Host` this should no longer be possible. We can create an issue and address this in a followup. I'd be willing to take that task if we agree.

**Review by mwear — APPROVED** [review]
LGTM. I have one non-actionable comment and a suggestion for a followup.

File: service/service.go
Comment (TylerHelmuth) [reply service/service.go:199]: I've captured this in https://github.com/open-telemetry/opentelemetry-collector/issues/10792.  This PR is already complicated enough I think that can be done in a followup.

File: component/telemetry.go
Comment (Wise-Wizard) [reply component/telemetry.go:38]: Yes sure, go ahead. It will actually be better to use componentstatus.Event over the previous implementation.

**Comment by mx-psi** [comment]
```
--- FAIL: Test_ComponentStatusReporting_SharedInstance (10.01s)
    testing.go:1398: race detected during execution of test
```

**Review by mx-psi — DISMISSED** [review]
Blocked based on discussion on the 2024-07-08 Collector stability WG meeting until after v0.107.0 is released

**Comment by mx-psi** [comment]
@TylerHelmuth Can you fix the merge conflicts?

**Comment by mx-psi** [comment]
@TylerHelmuth How do you plan on handling contrib tests failure? Is there any way to make them pass before merging this PR?

**Review by mx-psi — APPROVED** [review]
LGTM. As discussed on the 2024-08-14 Collector Stability WG meeting I will merge this by EOW if there are no further comments

**Comment by TylerHelmuth** [comment]
@mwear and I will handle the breaking changes via a `make update-otel` PR. For the components that are reporting fatal error it'll be a 1 line change.  For healthcheck extension v2 it'll be a lot of changes.

We'll also end up duplicating the logic from core's `sharedcomponent` into contrib's `sharedcomponent`.

**Comment by mwear** [comment]
> @mwear and I will handle the breaking changes via a `make update-otel` PR. For the components that are reporting fatal error it'll be a 1 line change. For healthcheck extension v2 it'll be a lot of changes.
> 
> We'll also end up duplicating the logic from core's `sharedcomponent` into contrib's `sharedcomponent`.

I'll handle healthcheckv2extension. We can divide up anything else that remains.

**Pull request merged**
